### ISSUE TITLE
Relations need ability to manually overriding the class, table, foreign_key

### ISSUE THREAD

**Issue by robacarp** [amberframework/granite#103]
Somewhat pursuant to #76, I'd like to add overrides for the has_many and belongs_to macros which allow a syntax like this:

```crystal
class IPAddress
  belongs_to :host
end
```

```crystal
class Host
  has_many :ip_addresses, class: IPAddress
end
```

The inability of any sort of inflector to correctly singularize "ip_addresses" means it correctly tries to look for a table called "ip_addresses" but tries to build them into a class called `IPAddresse`. One solution to this is to implement a proper inflector, but there are always going to be failure cases where "proper english" and "what people actually want to do" don't match up (or they're programming in one of the other myriad of non-english languages).

Providing the ability to override the generated class name would help, provide immediate benefit where pluralization by adding and removing "s" fails, and make model code more readable.

Providing the ability to override the foreign_key and type would also address #92 and #93 as well and would be similar work.

This prototype code is _mostly working_ but will need more refinement before it can be viable. @amberframework/granite-orm-contributors thoughts?

```crystal
  macro has_many(
    relation_plural, class relation_class_name = nil, table_name relation_table_name = nil,
    through join_relation = nil, through_table_name join_table_name = nil)
    {%
      relation_plural = relation_plural.id
      relation_singular = relation_plural[0...-1]

      name_space = @type.name.gsub(/::/, "_").underscore.id
      self_plural = SETTINGS[:table_name] || name_space + "s"
      self_singular = self_plural[0...-1]

      if relation_class_name == nil
        relation_class_name = relation_singular.camelcase
      end

      if relation_table_name == nil
        relation_table_name = relation_class_name.id.underscore + "s"
      end

      if join_table_name == nil
        join_table_name = join_relation
      end

      foreign_key = "#{relation_table_name.id}.#{self_singular}_id"
    %}

    def {{ relation_plural.id }} : Array({{ relation_class_name }})
      return [] of {{ relation_class_name }} unless id

      {% if join_relation %}
        query = <<-SQL
          JOIN {{ join_table_name.id }} ON {{ join_table_name.id }}.{{ relation_singular }}_id = {{ relation_plural }}.id
          WHERE {{ join_table_name.id }}.{{ name_space.id }}_id = ?
        SQL
        {{ relation_class_name }}.all(query, [id])
      {% else %}
        query = "WHERE {{ foreign_key.id }} = ?"
        {{ relation_class_name }}.all(query, [id])
      {% end %}
    end
```

**Comment by drujensen** [comment]
@robacarp Can you create a PR for this?

**Comment by robacarp** [comment]
@drujensen Thanks for looking! I'd love to, but I believe I've come across something that needs to be decided before proceeding. At least, I don't feel I should be the sole decision maker for this problem.

The rails behavior around `has_many through:` is divergent to the granite behavior. I'm not _sure_ the right way to resolve it. Rails reflects on the type of association it's navigating `through` and changes its query to match. The implementation I have been working on [over here](https://github.com/amberframework/granite-orm/compare/master...robacarp:association_specific_names) is as far as I've gotten, and I stumbled upon this problem while maniplating the test models for `has_many through`.

In short, there are two different ways something can be "through":

- "has an belongs to many" style (with an explicit join table) and
- "has many that has many" style

Rails inspects and delivers the appropriate query. As currently coded, Amber cannot do this. I believe that choosing the right thing to do here is more complicated than only implementing one of these, because we're following in footsteps of Rails relationships enough here that I expect people to come in with big expectations.

I see three ways to resolve this, roughly in order of complexity:

- Document that `has_many through:` doesn't support "has many that has many" or any other relations besides the one relational style it does support (I don't like this idea, at least not beyond the short term).
- Completely rework relationships into a macro datastructure to be computed and rendered similarly to how fields work and render `to_json` etc. This may provide the means to reflect adequately and just do the right thing.
- Come up with a completely new DSL of relating models to each other, which is more adequate than the sometimes cumbersome `belongs_to` and `has_many` verbiage. I'm not great with names, but I'm sure other ORMs out there have good examples to follow.

**Comment by drujensen** [comment]
@robacarp I vote for option 1

One of the major goals I have for Granite is to keep it simple.  This is why I'm avoiding AREL like DSL for SQL.  Also, I am hoping to keep the macros something that you could easily replace with your own methods.  Nothing that couldn't be done in a couple lines of code.  I want the macros to be simple helpers, nothing more.

One goal that you mentioned above is compatibility with Rails, but this is not one of Amber's goals.  We want to take the best ideas and concepts from other frameworks but leave out ones that we think are not so good.

My vote is that we never support "has many that has many".  I haven't ran into a situation where having a hidden join table is a good idea.  In Rails, I used to use the hidden join table, but almost always I had to change this to a model because the joining table needed properties.  I had some really bad designs in the past trying to get around the hidden join tables like putting fields in the wrong models.  

I prefer being opinionated on database design and help developers avoid bad decisions of which I consider this one of them.

**Comment by robacarp** [comment]
@drujensen I hear you. Let me clear something up. I don't think hidden join tables are a good idea, and my experience with the literal `has_and_belongs_to_many` rails provides is similarly sour. I think by using the jargon, I've confused the point (See also option 3).

Of the two situations I'm talking about, neither one relies on a hidden join table. However, they both rely on transparently joining across a middle table to provide the resultset.

ASCIIflow is a bit weak when it comes to ERD, but hopefully these diagrams add clarity and not confusion 😅 . I've provided a Table with a field and below the line are methods which should be provided when the models are related to each other.

```
+------------+   +-------------+   +------------+
|School      |   |County       |   |State       |
|  County Id |   |  State Id   |   |            |
+------------+   +-------------+   +------------+
|  #county   |   |  #state     |   |  #counties |
|  #state    |   |  #schools   |   |  #schools  |
+------------+   +-------------+   +------------+
```

Crucially, `State.schools` is a `has_many through:` relationship, but the foreign keys aren't centrally located on a single join table.

```
+---------------+       +--------------------+
|Student        |       |Enrollment          |
|               |       |  class_id          |
+---------------+       |  student_id        |
| #classes      |       +--------------------+
| #teachers     |       |  #class            |
| #enrollments  |       |  #student          |
+---------------+       +--------------------+


+-------------------+    +-------------------+
|Teacher            |    |Class              |
|                   |    |  teacher_id       |
+-------------------+    +-------------------+
|  #students        |    |  #students        |
|  #classes         |    |  #enrollments     |
|  #enrollments     |    +-------------------+
+-------------------+
```

Here, `Student.classes` is a `has_many through:` where the foreign keys are centrally located. 
Even worse, `Student.teachers` is another `has_many through:` where the `through` is a `has_many through:`.

Rails provides this deep reflective association querying because it builds efficient queries out of these relationships, and I think the jargon `has_many` etc. will elicit expectations from people coming from Rails.

**Comment by drujensen** [comment]
@robacarp Thanks for clarifying.  Nice job adding diagrams!  :-)  

I was hesitant to add `belongs_to` and `has_many` because I knew it would open this can of worms.  I think my fears were justified.

Without context, it will be difficult to guess what the best query will be.  Should we eagerly load the models into the array?  Should we only load the first 100 rows?  What about the order clause? Is there a where clause or other more complex join needed?

I think trying to encapsulate all the possible use cases in the `has_many` macro is going to lead us to some very complex code and poorly performing queries.  Developers will forget to add indexes, eager load when not needed, N+1 queries, and other issues you constantly see in Rails apps.

For me, writing the `State.schools` method that performs a join query and returns a list is not that difficult to write.  It will also give me full control over the query and caching mechanism.  I can determine what should be loaded into memory or cached long term, what the paging mechanism should be, how they should be ordered, what the where clause should be, what the indexes should be, etc.

This is where I start to ask, "is it worth it?"

**Comment by robacarp** [comment]
I think perhaps what we're demonstrating in this conversation here is that there needs to be a separation of concerns, and perhaps `belongs_to` and friends should live somewhere else?

ActiveRecord, Doctrine, etc all provide an abstraction layer around SQL that makes it easy to build an MVP, but also allow for tuning and optimizing later. Sometimes it's subtly risky to build a scaled application around those paradigms, and without attention to detail `n+1` queries are running etc. But I think the `ready to MVP` attitude is where Rails gets it's majesty, and common usage.

I guess what I'm saying is, building any one method which queries the database and builds and returns associated objects is straightforward. But building and maintaining the [100s+ of models and 1000s of relationships](https://github.com/tootsuite/mastodon/tree/161c72d66d25bb8f5ff492e36a8521c701ab8afc/app/models) manually which inevitably inhabit a mature application becomes a significant technical ownership. Relationships in an application scale at best around `n^2`, and at worst perhaps even `2^n` (n = number of models). 

What's really powerful about AREL is the ability to generate partial queries, and patch onto them elsewhere.  The ability to say e.g. `State.schools` and have it order them by `id ASC` by default because postgres effectively _randomizes_ rows without an order clause is important. But **more important** than that is the ability to say `State.schools.order_by :built_on, DESC` and get what you expect, instead of a query with two `ORDER` clauses. 

> This is where I start to ask, "is it worth it?"

I agree, but I might ask the question slightly different. Is this where this code should live? The reason an ORM exists is to "dry up" database code. Writing the same logic with subtle differences between models is what the dynamic web was before ORMs, and it was terrible. Select some rows, build objects out of those rows, select rows in another table with yet another `WHERE id IN(#{list_of_ids.join})` clause, ad nauseum.

**Comment by drujensen** [comment]
>> Perhaps belongs_to and friends should live somewhere else?

Perhaps.

>> But I think the ready to MVP attitude is where Rails gets it's majesty, and common usage.

Yes, however, its a double edged sword.  I think there is a balance between rapid development and production readiness that we need to try and find.

>> But building and maintaining the 100s+ of models and 1000s of relationships manually which inevitably inhabit a mature application becomes a significant technical ownership.

Yes, true. This is where you would refactor and write common base class methods.  If we can find some common methods or macros i.e. `belongs_to` that can be leveraged by the community, we should look at bringing them into the fold. I'm not against this.  

However, we should focus on the most common use cases and be careful not to write a whole other layer of complexity to handle every use-case.

>> What's really powerful about AREL is the ability to generate partial queries, and patch onto them elsewhere.

Yes, there are some niceties with AREL for MVP but I don't think its worth it.  I find it makes it difficult to figure out what query is generated and ran, what indexes need to be added to optimize performance, etc.

I used to do consulting work and just go around and fix Rails App performance issues.  80% of the work was just adding missing indexes.  It was quite sad that Rails Apps were running so slow after the site had a couple hundred entries just because the developers didn't know how to optimize the DB.

Should I blame AREL and ActiveRecord for this?  I don't know.  I just know that the Rails developers didn't seem to understand how the DB worked.  I think the abstraction created a slew of developers that never learned basic [ACID](https://en.wikipedia.org/wiki/ACID) principles.

What I found even more ironic is that the community started moving the NoSQL solutions (i.e. MongoDB) thinking this is why their apps were slow. They replaced one set of problems with another (data integrity). 

I digress.

**Comment by robacarp** [comment]
>  I think there is a balance between rapid development and production readiness that we need to try and find.

@drujensen I think we agree.

I have a hard time knowing how and where to draw that line. The Jennifer ORM may well swallow the bulk of the expectations that ActiveRecord users would come to the table with, but I was dissuaded by their query syntax. If it doesn't, undoubtedly something will emerge. 

Thanks for a good discussion on project perspective and design goals. I don't really know how to move forward.

I've toyed with proposing a completely different language for talking about database relationships, using generated and committed methods instead of macros, and a few other ideas and I don't have confidence in any of it to say it's worth pushing out.

**Comment by drujensen** [comment]
@robacarp Thanks for your help here.  I hope I didn't discourage you too much.  Hopefully we can find that balance.

What do you think of a `query-builder` that can be used with granite? https://github.com/izniburak/query-builder

**Comment by Blacksmoke16** [comment]
This is still relevent.  As of now associations assume there the PK is `id` with FK is like `foreign_key_id`

**Comment by drujensen** [comment]
fixed in #278
